### PR TITLE
Explore - UI panels state persistance in url

### PR DESCRIPTION
**Pull request by dprokop**
~WIP.~

In 2ddccb4a214a1828bde0ffb3c0d0191773d8146a I have temporarily disabled the check whether or not panel is collapsed to perform queries. 

Right now panel's UI state recovery from URL is tricky, cause when all DS's supported panels are collapsed, the start page is shown instead of collapsed panels. That's because the check is performed in `runQueries` action that's being dispatched from `loadDatasource` action. 

As a next step I want to:
- remove `runQueries` execution from `loadDatasource` and dispatch it independently in related actions after datasource is loaded.
- add `ignoreUIState` property to `runQueries` action payload, to allow queries execution when initialising Explore. 

See below for update

Any thoughts @torkelo @hugohaggmark ?

### CODE REVIEW

**Comment by torkelo** [comment]
looks good. 

> As a next step I want to:
> remove runQueries execution from loadDatasource and dispatch it independently in related action
> after datasource is loaded.
> add ignoreUIState property to runQueries action payload, to allow queries execution when initialising 
> Explore.

Sounds like a good plan. 

hope this does not inflict to much merge issues with the refactoring your working on @hugohaggmark

File: public/app/core/utils/explore.ts [hunk 1]
```diff
@@ -160,20 +169,42 @@ export function parseUrlState(initial: string | undefined): ExploreUrlState {
           to: parsed[1],
         };
         const datasource = parsed[2];
-        const queries = parsed.slice(3);
-        return { datasource, queries, range };
+        let queries = [],
+          ui;
```
Comment (hugohaggmark) [inline public/app/core/utils/explore.ts]: Maybe initialize this to something?

**Review by hugohaggmark — APPROVED** [review]

**Comment by dprokop** [comment]
Intoduced UI state segment to Explore's url.  To keep the urls simple I am basically adding `url:[showGraph, showLogs, showTable]`.

I think the same convention could be added to non-query related items, e.g. range and data source. Compact serialization that we use now is quite rigid. So the proposed url scheme could be:

`?[exploreId]=options:[rangeFrom, rangeTo, ...],queries:[query1, ...],ui:[showGraph,...]`

I'm gonna explore that when working on storing the dedup strategy in URL (#14445).

@hugohaggmark feel free to try merging this branch now.

File: public/app/features/explore/state/actions.ts [hunk 2]
```diff
@@ -572,7 +587,7 @@ export function removeQueryRow(exploreId: ExploreId, index: number): ThunkResult
 /**
  * Main action to run queries and dispatches sub-actions based on which result viewers are active
  */
-export function runQueries(exploreId: ExploreId) {
+export function runQueries(exploreId: ExploreId, ignoreUIState = false) {
```
Comment (torkelo) [inline public/app/features/explore/state/actions.ts]: not sure I understand this flag, when this relevant?

**Review by torkelo — COMMENTED** [review]
I did find one timing related issue with this, if open explore and write a prometheus query, get a graph & table. Then collapse the table & do a full page reload, the query for the table is still issued (but the table is correctly collapsed). 

Not sure if it's related but the way UI state works/is updated might make it very hard to support 
Browser back/forward https://github.com/grafana/grafana/issues/14301
It is supported for 90% of all url state actions in the dashboard (time range, variables, panel & dashboard view states). 

Not sure how we can accomplish that easily in explore with redux. It would involve listening to redux location updated actions and then parse UI state and trigger explore state update/action. But tricky to not let this cause infinite loops.

File: public/app/features/explore/state/actions.ts (hunk 2 above)
Comment (dprokop) [reply public/app/features/explore/state/actions.ts]: When requesting queries on Explore init. Whether or not the queries are issued depends on the state of the panels (which sort of makes sense). But, when entering Explore for the first time, I'm ignoring this, and issuing quieries anyway -> that's why the problem you have described below occurs. 

The queries are issued based on the `show[Graph|Logs|Table]` properties in state (and that's implemented in `runQueries` action), while they execution could have been issued from the panels (on mount / toggle). Gonna think what's there to be refactored to achive that.

**Pull request merged**